crates/price-feed/src/lib.rs

```diff
@@ -1,53 +1,135 @@
-use std::{collections::HashMap, sync::Arc};
+use std::{collections::HashMap, fmt::Display, sync::Arc};
 
+use alloy_primitives::Address;
 use fetching::fetch_price;
 use parking_lot::Mutex;
 pub use price::Price;
-#[cfg(test)]
 use rust_decimal::Decimal;
-use shielder_relayer::{PriceProvider, TokenInfo, TokenKind};
+use serde::{Deserialize, Serialize};
+use shielder_account::Token;
 use time::OffsetDateTime;
 use tokio::time::Duration;
 use tracing::warn;
+use utoipa::{schema, ToSchema};
 
-use crate::price_feed::price::Expiration;
+use crate::price::Expiration;
 
 mod fetching;
 mod price;
 
+#[derive(Copy, Clone, Debug, Default, Hash, Eq, PartialEq, Deserialize, Serialize, ToSchema)]
+pub enum TokenKind {
+    #[default]
+    Native,
+    ERC20 {
+        #[schema(value_type = String)]
+        address: Address,
+        decimals: u32,
+    },
+}
+
+impl Decimals for TokenKind {
+    fn decimals(&self) -> u32 {
+        match self {
+            TokenKind::Native => 18,
+            TokenKind::ERC20 { decimals, .. } => *decimals,
+        }
+    }
+}
+
+#[derive(Clone, Debug, Eq, PartialEq, Hash, Deserialize, Serialize)]
+pub struct SimpleKind {
+    pub name: String,
+    pub decimals: u32,
+}
+
+impl Decimals for SimpleKind {
+    fn decimals(&self) -> u32 {
+        self.decimals
+    }
+}
+
+impl From<TokenKind> for Token {
+    fn from(token_kind: TokenKind) -> Self {
+        match token_kind {
+            TokenKind::Native => Token::Native,
+            TokenKind::ERC20 { address, .. } => Token::ERC20(address),
+        }
+    }
+}
+
+impl Display for TokenKind {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        let str = match self {
+            TokenKind::Native => "native".to_string(),
+            TokenKind::ERC20 { address, .. } => format!("erc20:{address}"),
+        };
+        write!(f, "{str}")
+    }
+}
+
+#[derive(Clone, Debug, Eq, PartialEq, Deserialize, Serialize)]
+pub enum PriceProvider {
+    Dia(String),
+    Pyth(String),
+    Static(Decimal),
+}
+
+#[derive(Clone, Debug, Eq, PartialEq, Deserialize, Serialize)]
+pub struct TokenInfo<K> {
+    pub kind: K,
+    pub price_provider: PriceProvider,
+}
+
+pub trait Decimals {
+    fn decimals(&self) -> u32;
+}
+
+impl<K: Decimals> Decimals for TokenInfo<K> {
+    fn decimals(&self) -> u32 {
+        self.kind.decimals()
+    }
+}
+
 /// A collection of prices for various coins.
 ///
 /// The underlying structure is behind a mutex, and a process to update it
 /// asynchronously can be started with `start_price_feed`.
+///
+/// The type parameter `K` is the type used to identify tokens.
+/// Apart from being `Clone`, `Eq`, `PartialEq` and `Hash`, it must also implement the `Decimals` trait
+/// to provide the number of decimals for each token. `TokenKind` is a legacy implementation that works
```

Comment:
Maybe to "enforce" this assumption in client libs, we can add following type aliases in the new crate
```
pub type LegacyTokenInfo = TokenInfo<TokenKind>;
pub type SimpleTokenInfo = TokenInfo<SimpleKind>;
```
And use those symbols in the the clients of the `price-feed` crate.